soc/xtensa/intel_adsp/tools/cavstool_client.py

```diff
@@ -41,27 +44,47 @@ def send_cmd(self, cmd):
 
             if ack == CMD_LOG_START:
                 self.monitor_log()
-            elif ack == CMD_LOG_STOP:
-                log.info(f"Stop output.")
             elif ack == CMD_DOWNLOAD:
                 self.run()
             else:
                 log.error(f"Receive incorrect msg:{ack} expect:{cmd}")
 
-    def download(self, filename):
+    def uploading(self, filename):
         # Send the FW to server
+        fname = os.path.basename(filename)
+        fsize = os.path.getsize(filename)
+
+        md5_tx = hashlib.md5(open(filename,'rb').read()).hexdigest()
+
+        # Pack the header and the expecting packed size is 78 bytes.
+        # The header by convention includes:
+        # size(4), filename(42), MD5(32)
+        values = (fsize, fname.encode('utf-8'), md5_tx.encode('utf-8'))
+        log.info(f'filename:{fname}, size:{fsize}, md5:{md5_tx}')
+
+        s = struct.Struct(PACKET_HEADER_FORMAT_FW)
+        header_data = s.pack(*values)
+        header_size = s.size
+        log.info(f'header size: {header_size}')
+
         with open(filename,'rb') as f:
-            log.info('Sending...')
-            ret = self.sock.sendfile(f)
-            log.info(f"Done Sending ({ret}).")
+            log.info(f'Sending...')
+
+            total = self.sock.send(header_data)
+            total += self.sock.sendfile(f)
+
+            log.info(f"Done Sending ({total}).")
+
+            rck = self.sock.recv(BUF_SIZE).decode("utf-8")
```

Comment:
Again you should not read `BUF_SIZE` but OK, here it's less likely to fail.

Reply:
ok, fixed.